### DISCUSSION TITLE
Containers vs Serverless for the future

### DISCUSSION THREAD

**Discussion by 3plcentraldev** [BretFisher/ama#201] [Q&A]
Hello Bret, I have taken your course and saw you speak recently in southern utah.  As a big fan and the person that got me really introduced to using docker, I would be interested in your thoughts in regards regards to containers vs serverless.  Some individuals see them as competing fields, some as complimentary technologie, some as mutually exclusive.  Would be interested to hear your thoughts especially in regards to where you think this is heading in the future.

**Comment by BretFisher** [comment]
This could be a single paragraph answer or a series of blog posts. 😬

In short, it's not either-or. They do not compete, they compliment. 

Since there is not a single definition of "serverless" I'll just discuss Functions as a Service (FaaS), which is where I think most of the current solutions will end up.

* If you are doing greenfield development, **and** you are using one of the few languages that Functions as a Service supports, **and** you've worked out all the decisions on *which* FaaS platform you want, then FaaS sounds good.
* Everything else should be in a container.

We've got 50 years of software development that can all run in containers, usually without any code changes. FaaS requires new development, just like going from monolith to microservice.

I really like the idea of FaaS running in containers, because uploading zip files of my code and dependencies is a "solved problem" with containers. Last I checked the Lamda DevOps workflow it sucked.  Some think containers are too "heavy" but IMO they meet all the requirements of deploying/running functions, and a lot of the current FaaS platforms let you deploy your code as containers. See [OpenFaaS](https://github.com/openfaas/faas)

To me, FaaS is just an architecture decision that most code today won't fit into yet. It's like saying "microservices or containers". That's a silly comparison. One is an implementation detail (containers) and one is an architecture pattern (microservices). The same with FaaS. It's an architecture pattern. You can do it in containers, or zip files, or something else.

In the end, we'll all likely be using old monoliths, some microservices, and some FaaS... together, to form a single set of customer services... but I bet they'll all be in containers.

See some Kelsey Hightower tweets for more thoughts:

https://twitter.com/kelseyhightower/status/968146259188973568
https://twitter.com/kelseyhightower/status/921556474161111040
https://twitter.com/kelseyhightower/status/902525728725405697
https://twitter.com/kelseyhightower/status/1011368682960777216

**Comment by simonsouv** [comment]
Hi all,

Thank you Bret for your valuable inputs

Simon
